// @keix/message-store-client/dist/endpoint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
exports.__esModule = true;
exports.createEndpoint = void 0;
var express_1 = __importDefault(require("express"));
var clientVersion = require("../package.json").version;
var endpointPort = (_a = process.env.ENDPOINT_PORT) !== null && _a !== void 0 ? _a : 8080;
function createEndpoint() {
    var app = express_1["default"]();
    var startDate = new Date();
    var keepAlive = {};
    var stats = {};
    app.get("/info", function (req, res) {
        var uptime = new Date().getTime() - startDate.getTime();
        res.json({
            clientVersion: clientVersion,
            uptime: uptime,
            stats: stats,
            keepAlive: keepAlive
        });
    });
    return {
        onKeepAlive: function (subscriberId, date) {
            keepAlive[subscriberId] = date;
        },
        onStatsUpdated: function (subscriberId, stat) {
            stats[subscriberId] = stat;
        },
        listen: function () {
            app.listen(endpointPort, function () {
                console.log("Message store internal endpoint running at " + endpointPort + "...");
            });
        }
    };
}
exports.createEndpoint = createEndpoint;
